mempool/priority/mempool.go

```diff
@@ -0,0 +1,809 @@
+	// semantics of the Mempool interface require the caller to hold it, and we
+	// can't change that without disrupting existing use.
+	txmp.mtx.Unlock()
+	defer txmp.mtx.Lock()
+
+	return txmp.proxyAppConn.Flush(context.TODO())
+}
+
+// EnableTxsAvailable enables the mempool to trigger events when transactions
+// are available on a block by block basis.
+func (txmp *TxMempool) EnableTxsAvailable() {
+	txmp.mtx.Lock()
+	defer txmp.mtx.Unlock()
+
+	txmp.txsAvailable = make(chan struct{}, 1)
+}
+
+// TxsAvailable returns a channel which fires once for every height, and only
+// when transactions are available in the mempool. It is thread-safe.
+func (txmp *TxMempool) TxsAvailable() <-chan struct{} { return txmp.txsAvailable }
+
+// CheckTx adds the given transaction to the mempool if it fits and passes the
+// application's ABCI CheckTx method.
+//
+// CheckTx reports an error without adding tx if:
+//
+// - The size of tx exceeds the configured maximum transaction size.
+// - The pre-check hook is defined and reports an error for tx.
+// - The transaction already exists in the cache.
+// - The proxy connection to the application fails.
+//
+// If tx passes all of the above conditions, it is passed (asynchronously) to
+// the application's ABCI CheckTx method and this CheckTx method returns nil.
+// If cb != nil, it is called when the ABCI request completes to report the
+// application response.
+//
+// If the application accepts the transaction and the mempool is full, the
+// mempool evicts one or more of the lowest-priority transaction whose priority
+// is (strictly) lower than the priority of tx and whose size together exceeds
+// the size of tx, and adds tx instead. If no such transactions exist, tx is
+// discarded.
+func (txmp *TxMempool) CheckTx(
+	tx types.Tx,
+	cb func(*abci.ResponseCheckTx),
+	txInfo mempool.TxInfo,
+) error {
+	// During the initial phase of CheckTx, we do not need to modify any state.
+
+	// Reject transactions in excess of the configured maximum transaction size.
+	if len(tx) > txmp.config.MaxTxBytes {
+		return mempool.ErrTxTooLarge{Max: txmp.config.MaxTxBytes, Actual: len(tx)}
+	}
+
+	// If a precheck hook is defined, call it before invoking the application.
+	if err := txmp.preCheck(tx); err != nil {
+		txmp.metrics.FailedTxs.Add(1)
+		return mempool.ErrPreCheck{Err: err}
+	}
+
+	// Early exit if the proxy connection has an error.
+	if err := txmp.proxyAppConn.Error(); err != nil {
+		return err
+	}
+
+	txKey := tx.Key()
+
+	// Check for the transaction in the cache.
+	if !txmp.cache.Push(tx) {
+		// If the cached transaction is also in the pool, record its sender.
+		if elt, ok := txmp.txByKey[txKey]; ok {
+			txmp.metrics.AlreadySeenTxs.Add(1)
+			w := elt.Value.(*WrappedTx)
+			w.SetPeer(txInfo.SenderID)
+		}
+		return mempool.ErrTxInCache
+	}
+
+	// At this point, we need to ensure that passing CheckTx and adding to
+	// the mempool is atomic.
+	txmp.Lock()
+	defer txmp.Unlock()
```

Comment:
[blocking] these lines need to be lifted above to before line 205 to fix a data race

Ref: https://github.com/celestiaorg/celestia-core/pull/1655/files